File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char>map;
        int left=0;
        int right=0;
        int maxlen=0;
        while(right<s.length())
        {
            if(!map.contains(s[right]))
            {
                map.insert(s[right]);
                maxlen=max(maxlen,right-left+1);
                right++;
            }
            else
            {
                map.erase(s[left]);
                left++;
            }
        }
        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        HashMap<Character,Integer>mp=new HashMap<>();

        int l=0;
       
        int maxlen=0;
        for(int r=0;r<n;r++)
        {
            char c=s.charAt(r);
            if(mp.containsKey(c))
            {
                l=Math.max(l,mp.get(c)+1);
            }
            mp.put(c,r);
            maxlen=Math.max(maxlen,r-l+1);
            
        }
        return maxlen;
    }
}